WIP: Initial Program creation / list views.

jimabramson opened the pull request:
ECOM-2186

@clintonb @rlucioni  I'm working on the authentication / roles part of the ACs next, but ready for feedback on what's implemented so far.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
+
+
+# TODO complete model docstring
+# TODO internationalize help text?   Or get rid of it!
```
Internationalize.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
+        blank=True,
+    )
+
+    category = models.CharField(
```
Looking forward, what are the other categories/types of programs?

jimabramson replied:
None.  This is somewhat provisional; since xseries will be an edx-specific category this should be made a setting or plugin (using stevedore or similar).  For now I'm keeping it simple.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
+
+    category = models.CharField(
+        help_text='The category / type of Program.',
+        max_length=32,
```
Why the limitation? Is this user-facing?

jimabramson replied:
Quasi-arbitrary, but I think it's good practice to define a max length for UI/presentation sanity.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
+    )
+
+    certificate_type = models.CharField(
+        help_text='Optional certification criteria for Course Runs associated with this Program.',
```
Lowercase course runs.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
+    certificate_type = models.CharField(
+        help_text='Optional certification criteria for Course Runs associated with this Program.',
+        max_length=32,
+        choices=["None", None] + _choices(constants.CertificateType.VERIFIED),
```
Mark `"None"` for translation.

clintonb commented on programs/apps/api/models.py:
```diff
@@ -1,4 +1,65 @@
-# should be created here. As the API evolves, models may become more
-# specific to a particular version of the API. In this case, the models
-# in question should be moved to versioned sub-package.
+"""
```
I should probably get rid of `api/models.py`. The odds of there being API-specific are pretty low. This file belongs in `apps/programs/models.py`, where `apps/programs` contains the logic for Programs.

clintonb commented on programs/apps/api/serializers.py:
```diff
@@ -1,4 +1,30 @@
+class ProgramSerializer(serializers.ModelSerializer):
+    """General-purpose serializer for the Program model."""
+
+    class Meta:  # pylint: disable=missing-docstring, old-style-class
```
Use `class Meta (object)` to get around the pylint warning.

clintonb commented on programs/apps/api/serializers.py:
```diff
@@ -1,4 +1,30 @@
+
+    def validate_status(self, status):
+        """
+        Prevent creation of Programs with a status other than UNPUBLISHED.
```
What about editing? Is this limitation/validation temporary?

jimabramson replied:
unpublished is the only valid initial state.  transitions to successive states are expected to happen after Programs are fully built up, which is out of scope for this initial set of functionality.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+STATUSES = (ProgramStatus.UNPUBLISHED, ProgramStatus.ACTIVE, ProgramStatus.RETIRED, ProgramStatus.DELETED)
+
+
+class ProgramFactory(factory.django.DjangoModelFactory):  # pylint: disable=missing-docstring
```
Move to `apps/programs/factories.py`.

jimabramson replied:
OK, gonna make that apps/programs/tests/factories.py tho.

clintonb commented:
`ddt` needs to be added to `requirements/test.txt`.

clintonb commented on programs/apps/api/v1/urls.py:
```diff
@@ -1,3 +1,11 @@
+# Additionally, we include login URLs for the browsable API.
+urlpatterns = [
+    url(r'^programs/', views.ProgramsView.as_view(), name='programs-list'),
+    # url(r'^api-auth/', include('rest_framework.urls', namespace='rest_framework'))
```
Delete.

clintonb commented on programs/apps/api/v1/views.py:
```diff
@@ -1 +1,17 @@
+
+
+# TODO complete docstring
+class ProgramsView(generics.ListCreateAPIView):
```
`class ProgramsView(mixins.CreateModelMixin, mixins.ListModelMixin,  viewsets.GenericViewSet)`

I'd like to get in the habit of using [`Viewsets`](http://www.django-rest-framework.org/api-guide/viewsets/) as they clean up some boiler plate view and url code. Also, if/when we later add the retrieve/delete actions, we need only add a mixin and tests.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+            self.assertIn("not a valid choice", content["certificate_type"][0])
+
+    @ddt.data(None, "", ProgramCategory.XSERIES, "unrecognized")
+    def test_create_with_category(self, category):
```
Split this into two tests: one for the accepted category, another for the unacceptable values. That should help clean up the if-statements/branching.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+            self.assertIn("not a valid choice", content["status"][0])
+
+    @ddt.data(None, "", CertificateType.VERIFIED, "unrecognized")
+    def test_create_with_certificate_type(self, certificate_type):
```
Same comment as `test_create_with_category`: split into two tests.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+    Tests for listing / creating Programs.
+    """
+
+    def setUp(self):
```
Delete.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+        self.assertNotIn(ProgramStatus.DELETED, set(obj["status"] for obj in content))
+
+    def test_create(self):
+        """
```
Verify the API supports the creation of Programs.

clintonb commented on programs/apps/api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,159 @@
+            expected_status = 400
+
+        response = self.client.post(reverse("api:v1:programs-list"), data=data)
+        print field, response.content
```
Delete.

rlucioni commented on programs/apps/api/serializers.py:
```diff
@@ -1,4 +1,32 @@
+        Prevent creation of Programs with a status other than UNPUBLISHED.
+        """
+        if self.instance is None and status != constants.ProgramStatus.UNPUBLISHED:
+            msg = _("When creating a Program, \"{}\" is not a valid choice.")
```
To make this easier on translators, use a keyword as a placeholder. Also, you can print literal curly braces by doubling them up:
```python
msg = _("When creating a Program, {{status}} is not a valid choice.")
```

jimabramson commented:
@rlucioni addressed your comment.

@rlucioni @clintonb i'd like to go ahead and merge this, then set up a separate follow-up PR for auth functionality.  If you're down, give me your comments or thumbs up...

clintonb commented on programs/apps/api/serializers.py:
```diff
@@ -1,4 +1,32 @@
+from rest_framework import serializers
+
+
+# TODO add back URLs
```
Does this still need to be done?

clintonb commented on programs/apps/programs/constants.py:
```diff
@@ -0,0 +1,24 @@
+"""
```
Use `from __future__ import unicode_literals` to avoid having to mark all strings as Unicode.

clintonb commented on programs/apps/programs/tests/factories.py:
```diff
@@ -0,0 +1,18 @@
+
+
+class ProgramFactory(factory.django.DjangoModelFactory):  # pylint: disable=missing-docstring
+    class Meta:  # pylint: disable=missing-docstring, old-style-class
```
Use `class Meta(object)` to get rid of the pylint old-style-class violation.

clintonb commented:
:+1:

rlucioni commented:
:+1: once commits are squashed.
